uaclient/tests/test_pip.py

```diff
@@ -0,0 +1,118 @@
+        "file_content,expected",
+        (
+            (
+                "",
+                dedent(
+                    """\
+                [global]
+                index-url = http://bearer:token@python.esm.ubuntu.com/simple
+                index = http://bearer:token@python.esm.ubuntu.com/simple
+                """
+                ),
+            ),
+            (
+                dedent(
+                    """\
+                [freeze]
+                timeout = 10
+                """
+                ),
+                dedent(
+                    """\
+                [freeze]
+                timeout = 10
+
+                [global]
+                index-url = http://bearer:token@python.esm.ubuntu.com/simple
+                index = http://bearer:token@python.esm.ubuntu.com/simple
+                """
+                ),
+            ),
+            (
+                dedent(
+                    """\
+                [global]
+                index-url = www.mypip.com
+                """
+                ),
+                dedent(
+                    """\
+                [global]
+                index-url = http://bearer:token@python.esm.ubuntu.com/simple
+                index = http://bearer:token@python.esm.ubuntu.com/simple
+                """
+                ),
+            ),
+            (
+                dedent(
+                    """\
+                [global]
+                index-url = www.mypip.com
+
+                [freeze]
+                timeout = 10
+                """
+                ),
+                dedent(
+                    """\
+                [global]
+                index-url = http://bearer:token@python.esm.ubuntu.com/simple
+                index = http://bearer:token@python.esm.ubuntu.com/simple
+
+                [freeze]
+                timeout = 10
+                """
+                ),
+            ),
+            (
+                None,
+                dedent(
+                    """\
+                [global]
+                index-url = http://bearer:token@python.esm.ubuntu.com/simple
+                index = http://bearer:token@python.esm.ubuntu.com/simple
+                """
+                ),
+            ),
+        ),
+    )
+    def test_update_pip_conf(self, tmpdir, file_content, expected):
+        file_path = tmpdir / "pip.conf"
+
+        if file_content:
+            with file_path.open("w") as f:
+                f.write(file_content)
+
+        with mock.patch("uaclient.pip.PIP_CONFIG_FILE", file_path.strpath):
+            update_pip_conf(self.test_config_dict())
+
+        with file_path.open("r") as f:
+            """
+            We are sorting the result because we cannot easily allow
+            configparser to write the config in a sorted manner, meaning
+            that the order of the lines may not match if compared directly
+            """
+            assert sorted(f.read().strip()) == sorted(expected.strip())
```

Comment:
trying to resort the content of the confg file doesn't tell us that each section has the desired content. 

Let's work with dictionaries for this to make sure each section matches.
 Here's a patch of what I was thinking 

```diff
diff --git a/uaclient/tests/test_pip.py b/uaclient/tests/test_pip.py
index dcefab9..4b7a16a 100644
--- a/uaclient/tests/test_pip.py
+++ b/uaclient/tests/test_pip.py
@@ -1,5 +1,6 @@
 """Tests related to uaclient.pip module."""
 
+from configparser import ConfigParser
 import mock
 import pytest
 
@@ -20,18 +21,23 @@ class TestPipConfUpdate:
 
         return {"global": {"index-url": index_url, "index": index}}
 
+    def _cfg_to_dict(self, cfg_file):
+        """Return a ConfigParser dict representation of a config file."""
+        cfg_parser = ConfigParser()
+        cfg_parser.read(cfg_file)
+        cfg_dict = {}
+        for s in cfg_parser.sections():
+            cfg_dict[s] = {}
+            for option in cfg_parser[s]:
+                cfg_dict[s][option] = cfg_parser[s][option]
+        return cfg_dict
+
     @pytest.mark.parametrize(
         "file_content,expected",
         (
             (
                 "",
-                dedent(
-                    """\
-                [global]
-                index-url = http://bearer:token@python.esm.ubuntu.com/simple
-                index = http://bearer:token@python.esm.ubuntu.com/simple
-                """
-                ),
+                {"global": {"index-url": "http://bearer:token@python.esm.ubuntu.com/simple", "index": "http://bearer:token@python.esm.ubuntu.com/simple"}}
             ),
             (
                 dedent(
@@ -40,16 +46,7 @@ class TestPipConfUpdate:
                 timeout = 10
                 """
                 ),
-                dedent(
-                    """\
-                [freeze]
-                timeout = 10
-
-                [global]
-                index-url = http://bearer:token@python.esm.ubuntu.com/simple
-                index = http://bearer:token@python.esm.ubuntu.com/simple
-                """
-                ),
+                {"freeze": {"timeout": "10"}, "global": {"index-url": "http://bearer:token@python.esm.ubuntu.com/simple", "index": "http://bearer:token@python.esm.ubuntu.com/simple"}}
             ),
             (
                 dedent(
@@ -58,13 +55,7 @@ class TestPipConfUpdate:
                 index-url = www.mypip.com
                 """
                 ),
-                dedent(
-                    """\
-                [global]
-                index-url = http://bearer:token@python.esm.ubuntu.com/simple
-                index = http://bearer:token@python.esm.ubuntu.com/simple
-                """
-                ),
+                {"global": {"index-url": "http://bearer:token@python.esm.ubuntu.com/simple", "index": "http://bearer:token@python.esm.ubuntu.com/simple"}}
             ),
             (
                 dedent(
@@ -76,43 +67,22 @@ class TestPipConfUpdate:
                 timeout = 10
                 """
                 ),
-                dedent(
-                    """\
-                [global]
-                index-url = http://bearer:token@python.esm.ubuntu.com/simple
-                index = http://bearer:token@python.esm.ubuntu.com/simple
-
-                [freeze]
-                timeout = 10
-                """
-                ),
+                {"freeze": {"timeout": "10"}, "global": {"index-url": "http://bearer:token@python.esm.ubuntu.com/simple", "index": "http://bearer:token@python.esm.ubuntu.com/simple"}}
             ),
             (
                 None,
-                dedent(
-                    """\
-                [global]
-                index-url = http://bearer:token@python.esm.ubuntu.com/simple
-                index = http://bearer:token@python.esm.ubuntu.com/simple
-                """
-                ),
+                {"global": {"index-url": "http://bearer:token@python.esm.ubuntu.com/simple", "index": "http://bearer:token@python.esm.ubuntu.com/simple"}}
             ),
         ),
     )
     def test_update_pip_conf(self, tmpdir, file_content, expected):
         file_path = tmpdir / "pip.conf"
 
-        if file_content:
-            with file_path.open("w") as f:
-                f.write(file_content)
+        if file_content is not None:
+
+            file_path.write(file_content)
 
         with mock.patch("uaclient.pip.PIP_CONFIG_FILE", file_path.strpath):
             update_pip_conf(self.test_config_dict())
 
-        with file_path.open("r") as f:
-            """
-            We are sorting the result because we cannot easily allow
-            configparser to write the config in a sorted manner, meaning
-            that the order of the lines may not match if compared directly
-            """
-            assert sorted(f.read().strip()) == sorted(expected.strip())
+        assert self._cfg_to_dict(file_path.strpath) == expected
```

Reply:
Yes, you are definitely right here, I missed the fact that the sections will not be taken in consideration. I have updated the code to use the approach you have provided. Thanks for that :)